File: iOS/Csound iOS Obj-C Examples/Csound iOS Examples/ViewControllers/BaseCsoundViewController.m
/*
 
 BaseCsoundViewController.m:
 
 Copyright (C) 2014 Steven Yi, Aurelius Prochazka
 Updated in 2017 by Dr. Richard Boulanger, Nikhil Singh
 
 This file is part of Csound iOS Examples.
 
 The Csound for iOS Library is free software; you can redistribute it
 and/or modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 2.1 of the License, or (at your option) any later version.   
 
 Csound is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU Lesser General Public License for more details.
 
 You should have received a copy of the GNU Lesser General Public
 License along with Csound; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
 02111-1307 USA
 
 */

#import "BaseCsoundViewController.h"

@interface BaseCsoundViewController ()
- (void)configureView;
@end

@implementation BaseCsoundViewController

#pragma mark - Managing the detail item

- (void)setDetailItem:(id)newDetailItem
{
    if (_detailItem != newDetailItem) {
        self.detailItem = newDetailItem;
        
        // Update the view.
        [self configureView];
    }
    
    if (self.masterPopoverController != nil) {
        [self.masterPopoverController dismissViewControllerAnimated:YES completion:nil];
    }        
}

- (void)configureView
{
    // Update the user interface for the detail item.
    
    if (self.detailItem) {
        self.detailDescriptionLabel.text = [self.detailItem description];
    }
}


-(void)viewWillDisappear:(BOOL)animated {
    
    [super viewWillDisappear:animated];
    
    [self.csound stop];
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
    // Return YES for supported orientations
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone) {
        return (interfaceOrientation != UIInterfaceOrientationPortraitUpsideDown);
    } else {
        return YES;
    }
}

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        self.title = NSLocalizedString(@"Detail", @"Detail");
    }
    return self;
}

-(UIModalPresentationStyle)adaptivePresentationStyleForPresentationController:(UIPresentationController *)controller {
    return UIModalPresentationNone;
}

-(UIModalPresentationStyle)adaptivePresentationStyleForPresentationController:(UIPresentationController *)controller traitCollection:(UITraitCollection *)traitCollection {
    return UIModalPresentationNone;
}

-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    [segue destinationViewController].modalPresentationStyle = UIModalPresentationPopover;
}

#pragma mark - Split view

- (void)splitViewController:(UISplitViewController *)svc willChangeToDisplayMode:(UISplitViewControllerDisplayMode)displayMode
{
    self.navigationItem.leftBarButtonItem = [self splitViewController].displayModeButtonItem;
    self.navigationItem.leftBarButtonItem.title = @"Csound for iOS";
    self.masterPopoverController = svc;
    
}

- (void)splitViewController:(UISplitViewController *)splitController willShowViewController:(UIViewController *)viewController invalidatingBarButtonItem:(UIBarButtonItem *)barButtonItem
{
    [self.navigationItem setLeftBarButtonItem:nil animated:YES];
    self.masterPopoverController = nil;
}

@end

File: iOS/Csound-for-iOS/Csound-iOS-Swift-Examples/Csound iOS SwiftExamples/BaseCsoundViewController.swift
/*
  
 BaseCsoundViewController.swift
 
 Nikhil Singh, Dr. Richard Boulanger
 Adapted from the Csound iOS Examples by Steven Yi and Victor Lazzarini
 
 This file is part of Csound iOS SwiftExamples.
 
 The Csound for iOS Library is free software; you can redistribute it
 and/or modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 2.1 of the License, or (at your option) any later version.
 
 Csound is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU Lesser General Public License for more details.
 
 You should have received a copy of the GNU Lesser General Public
 License along with Csound; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
 02111-1307 USA
 
 */

import UIKit

// Define base class for all examples
class BaseCsoundViewController: UIViewController {
    
    var infoText = ""
    var infoVC = UIViewController()
    var csound = CsoundObj()    // CsoundObj() property inherited by all examples
    
    var detailItem: AnyObject? {
        didSet(newDetailItem) {
            configureView()
            
            if masterPopoverController != nil {
                masterPopoverController?.dismiss(animated: true, completion: nil)
            }
        }
    }
    var detailDescriptionLabel: UILabel!
    var masterPopoverController: UIViewController?

    override func viewDidLoad() {
        if UIDevice.current.userInterfaceIdiom == .phone {
            self.navigationItem.leftBarButtonItem = splitViewController?.navigationItem.leftBarButtonItem
        }
        super.viewDidLoad()
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        csound.stop()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func configureView() {
        if detailItem != nil {
            detailDescriptionLabel.text = detailItem?.description
        }
    }
    
    // Method to display information in a popover, called from each inheriting class
    func displayInfo(_ sender: UIButton) {
        infoVC.modalPresentationStyle = .popover
        
        let infoTextView = UITextView(frame: CGRect(x: 0, y: 0, width: infoVC.preferredContentSize.width, height: infoVC.preferredContentSize.height))
        infoTextView.isEditable = false
        infoTextView.isSelectable = false

        infoTextView.attributedText = NSAttributedString(string: infoText)
        infoTextView.font = UIFont(name: "Menlo", size: 16)
        
        let popover = infoVC.popoverPresentationController
        infoVC.view.addSubview(infoTextView)
        popover?.delegate = self
        popover?.permittedArrowDirections = .up
        popover?.sourceView = sender
        popover?.sourceRect = sender.bounds
        present(infoVC, animated: true, completion: nil)
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        segue.destination.modalPresentationStyle = .popover
    }
    
    // MARK: Init
    required override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
}

// MARK: UISplitViewControllerDelegate
extension BaseCsoundViewController: UISplitViewControllerDelegate {
    func splitViewController(_ svc: UISplitViewController, willChangeTo displayMode: UISplitViewControllerDisplayMode) {
        navigationItem.leftBarButtonItem = splitViewController?.displayModeButtonItem;
        navigationItem.leftBarButtonItem?.title = "Csound for iOS"
        masterPopoverController = svc
    }
    
    func splitViewController(_ svc: UISplitViewController, willShow aViewController: UIViewController, invalidating barButtonItem: UIBarButtonItem) {
        navigationItem.leftBarButtonItem = nil
        masterPopoverController = nil
    }
}

// MARK: UIPopoverPresentationControllerDelegate
extension BaseCsoundViewController: UIPopoverPresentationControllerDelegate {
    func adaptivePresentationStyle(for controller: UIPresentationController) -> UIModalPresentationStyle {
        return .none    // Ensure infoVC presents as popover on iPhone
    }
    
    func adaptivePresentationStyle(for controller: UIPresentationController, traitCollection: UITraitCollection) -> UIModalPresentationStyle {
        return .none
    }
}
